File: backend/financeiro/src/model/fin_parcela_pagar_model.py
from src import db
from src.model.fin_status_parcela_model import FinStatusParcelaModel
from src.model.fin_tipo_pagamento_model import FinTipoPagamentoModel


class FinParcelaPagarModel(db.Model):
    __tablename__ = 'fin_parcela_pagar'

    id = db.Column(db.Integer, primary_key=True)
    numero_parcela = db.Column(db.Integer)
    data_emissao = db.Column(db.DateTime)
    data_vencimento = db.Column(db.DateTime)
    data_pagamento = db.Column(db.DateTime)
    desconto_ate = db.Column(db.DateTime)
    valor = db.Column(db.Float)
    taxa_juro = db.Column(db.Float)
    taxa_multa = db.Column(db.Float)
    taxa_desconto = db.Column(db.Float)
    valor_juro = db.Column(db.Float)
    valor_multa = db.Column(db.Float)
    valor_desconto = db.Column(db.Float)
    valor_pago = db.Column(db.Float)
    historico = db.Column(db.Text)
    id_fin_lancamento_pagar = db.Column(db.Integer, db.ForeignKey('fin_lancamento_pagar.id'))
    id_fin_status_parcela = db.Column(db.Integer, db.ForeignKey('fin_status_parcela.id'))
    id_fin_tipo_pagamento = db.Column(db.Integer, db.ForeignKey('fin_tipo_pagamento.id'))

    fin_status_parcela_model = db.relationship('FinStatusParcelaModel', foreign_keys=[id_fin_status_parcela])
    fin_tipo_pagamento_model = db.relationship('FinTipoPagamentoModel', foreign_keys=[id_fin_tipo_pagamento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_fin_lancamento_pagar = data.get('idFinLancamentoPagar')
        self.id_fin_cheque_emitido = data.get('idFinChequeEmitido')
        self.id_fin_status_parcela = data.get('idFinStatusParcela')
        self.id_fin_tipo_pagamento = data.get('idFinTipoPagamento')
        self.numero_parcela = data.get('numeroParcela')
        self.data_emissao = data.get('dataEmissao')
        self.data_vencimento = data.get('dataVencimento')
        self.data_pagamento = data.get('dataPagamento')
        self.desconto_ate = data.get('descontoAte')
        self.valor = data.get('valor')
        self.taxa_juro = data.get('taxaJuro')
        self.taxa_multa = data.get('taxaMulta')
        self.taxa_desconto = data.get('taxaDesconto')
        self.valor_juro = data.get('valorJuro')
        self.valor_multa = data.get('valorMulta')
        self.valor_desconto = data.get('valorDesconto')
        self.valor_pago = data.get('valorPago')
        self.historico = data.get('historico')

    def serialize(self):
        return {
            'id': self.id,
            'idFinLancamentoPagar': self.id_fin_lancamento_pagar,
            'idFinChequeEmitido': self.id_fin_cheque_emitido,
            'idFinStatusParcela': self.id_fin_status_parcela,
            'idFinTipoPagamento': self.id_fin_tipo_pagamento,
            'numeroParcela': self.numero_parcela,
            'dataEmissao': self.data_emissao.isoformat(),
            'dataVencimento': self.data_vencimento.isoformat(),
            'dataPagamento': self.data_pagamento.isoformat(),
            'descontoAte': self.desconto_ate.isoformat(),
            'valor': self.valor,
            'taxaJuro': self.taxa_juro,
            'taxaMulta': self.taxa_multa,
            'taxaDesconto': self.taxa_desconto,
            'valorJuro': self.valor_juro,
            'valorMulta': self.valor_multa,
            'valorDesconto': self.valor_desconto,
            'valorPago': self.valor_pago,
            'historico': self.historico,
            'finStatusParcelaModel': self.fin_status_parcela_model.serialize() if self.fin_status_parcela_model else None,
            'finTipoPagamentoModel': self.fin_tipo_pagamento_model.serialize() if self.fin_tipo_pagamento_model else None,
        }

File: frontend/financeiro/lib/app/data/model/fin_parcela_pagar_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:financeiro/app/infra/infra_imports.dart';
import 'package:financeiro/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class FinParcelaPagarModel {
	int? id;
	int? idFinLancamentoPagar;
	int? idFinChequeEmitido;
	int? idFinStatusParcela;
	int? idFinTipoPagamento;
	int? numeroParcela;
	DateTime? dataEmissao;
	DateTime? dataVencimento;
	DateTime? dataPagamento;
	DateTime? descontoAte;
	double? valor;
	double? taxaJuro;
	double? taxaMulta;
	double? taxaDesconto;
	double? valorJuro;
	double? valorMulta;
	double? valorDesconto;
	double? valorPago;
	String? historico;
	FinStatusParcelaModel? finStatusParcelaModel;
	FinTipoPagamentoModel? finTipoPagamentoModel;

	FinParcelaPagarModel({
		this.id,
		this.idFinLancamentoPagar,
		this.idFinChequeEmitido,
		this.idFinStatusParcela,
		this.idFinTipoPagamento,
		this.numeroParcela,
		this.dataEmissao,
		this.dataVencimento,
		this.dataPagamento,
		this.descontoAte,
		this.valor,
		this.taxaJuro,
		this.taxaMulta,
		this.taxaDesconto,
		this.valorJuro,
		this.valorMulta,
		this.valorDesconto,
		this.valorPago,
		this.historico,
		this.finStatusParcelaModel,
		this.finTipoPagamentoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero_parcela',
		'data_emissao',
		'data_vencimento',
		'data_pagamento',
		'desconto_ate',
		'valor',
		'taxa_juro',
		'taxa_multa',
		'taxa_desconto',
		'valor_juro',
		'valor_multa',
		'valor_desconto',
		'valor_pago',
		'historico',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero Parcela',
		'Data Emissao',
		'Data Vencimento',
		'Data Pagamento',
		'Desconto Ate',
		'Valor',
		'Taxa Juro',
		'Taxa Multa',
		'Taxa Desconto',
		'Valor Juro',
		'Valor Multa',
		'Valor Desconto',
		'Valor Pago',
		'Historico',
	];

	FinParcelaPagarModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idFinLancamentoPagar = jsonData['idFinLancamentoPagar'];
		idFinChequeEmitido = jsonData['idFinChequeEmitido'];
		idFinStatusParcela = jsonData['idFinStatusParcela'];
		idFinTipoPagamento = jsonData['idFinTipoPagamento'];
		numeroParcela = jsonData['numeroParcela'];
		dataEmissao = jsonData['dataEmissao'] != null ? DateTime.tryParse(jsonData['dataEmissao']) : null;
		dataVencimento = jsonData['dataVencimento'] != null ? DateTime.tryParse(jsonData['dataVencimento']) : null;
		dataPagamento = jsonData['dataPagamento'] != null ? DateTime.tryParse(jsonData['dataPagamento']) : null;
		descontoAte = jsonData['descontoAte'] != null ? DateTime.tryParse(jsonData['descontoAte']) : null;
		valor = jsonData['valor']?.toDouble();
		taxaJuro = jsonData['taxaJuro']?.toDouble();
		taxaMulta = jsonData['taxaMulta']?.toDouble();
		taxaDesconto = jsonData['taxaDesconto']?.toDouble();
		valorJuro = jsonData['valorJuro']?.toDouble();
		valorMulta = jsonData['valorMulta']?.toDouble();
		valorDesconto = jsonData['valorDesconto']?.toDouble();
		valorPago = jsonData['valorPago']?.toDouble();
		historico = jsonData['historico'];
		finStatusParcelaModel = jsonData['finStatusParcelaModel'] == null ? FinStatusParcelaModel() : FinStatusParcelaModel.fromJson(jsonData['finStatusParcelaModel']);
		finTipoPagamentoModel = jsonData['finTipoPagamentoModel'] == null ? FinTipoPagamentoModel() : FinTipoPagamentoModel.fromJson(jsonData['finTipoPagamentoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idFinLancamentoPagar'] = idFinLancamentoPagar != 0 ? idFinLancamentoPagar : null;
		jsonData['idFinChequeEmitido'] = idFinChequeEmitido != 0 ? idFinChequeEmitido : null;
		jsonData['idFinStatusParcela'] = idFinStatusParcela != 0 ? idFinStatusParcela : null;
		jsonData['idFinTipoPagamento'] = idFinTipoPagamento != 0 ? idFinTipoPagamento : null;
		jsonData['numeroParcela'] = numeroParcela;
		jsonData['dataEmissao'] = dataEmissao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissao!) : null;
		jsonData['dataVencimento'] = dataVencimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataVencimento!) : null;
		jsonData['dataPagamento'] = dataPagamento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataPagamento!) : null;
		jsonData['descontoAte'] = descontoAte != null ? DateFormat('yyyy-MM-ddT00:00:00').format(descontoAte!) : null;
		jsonData['valor'] = valor;
		jsonData['taxaJuro'] = taxaJuro;
		jsonData['taxaMulta'] = taxaMulta;
		jsonData['taxaDesconto'] = taxaDesconto;
		jsonData['valorJuro'] = valorJuro;
		jsonData['valorMulta'] = valorMulta;
		jsonData['valorDesconto'] = valorDesconto;
		jsonData['valorPago'] = valorPago;
		jsonData['historico'] = historico;
		jsonData['finStatusParcelaModel'] = finStatusParcelaModel?.toJson;
		jsonData['finTipoPagamentoModel'] = finTipoPagamentoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idFinLancamentoPagar = plutoRow.cells['idFinLancamentoPagar']?.value;
		idFinChequeEmitido = plutoRow.cells['idFinChequeEmitido']?.value;
		idFinStatusParcela = plutoRow.cells['idFinStatusParcela']?.value;
		idFinTipoPagamento = plutoRow.cells['idFinTipoPagamento']?.value;
		numeroParcela = plutoRow.cells['numeroParcela']?.value;
		dataEmissao = Util.stringToDate(plutoRow.cells['dataEmissao']?.value);
		dataVencimento = Util.stringToDate(plutoRow.cells['dataVencimento']?.value);
		dataPagamento = Util.stringToDate(plutoRow.cells['dataPagamento']?.value);
		descontoAte = Util.stringToDate(plutoRow.cells['descontoAte']?.value);
		valor = plutoRow.cells['valor']?.value?.toDouble();
		taxaJuro = plutoRow.cells['taxaJuro']?.value?.toDouble();
		taxaMulta = plutoRow.cells['taxaMulta']?.value?.toDouble();
		taxaDesconto = plutoRow.cells['taxaDesconto']?.value?.toDouble();
		valorJuro = plutoRow.cells['valorJuro']?.value?.toDouble();
		valorMulta = plutoRow.cells['valorMulta']?.value?.toDouble();
		valorDesconto = plutoRow.cells['valorDesconto']?.value?.toDouble();
		valorPago = plutoRow.cells['valorPago']?.value?.toDouble();
		historico = plutoRow.cells['historico']?.value;
		finStatusParcelaModel = FinStatusParcelaModel();
		finStatusParcelaModel?.descricao = plutoRow.cells['finStatusParcelaModel']?.value;
		finTipoPagamentoModel = FinTipoPagamentoModel();
		finTipoPagamentoModel?.descricao = plutoRow.cells['finTipoPagamentoModel']?.value;
	}	

	FinParcelaPagarModel clone() {
		return FinParcelaPagarModel(
			id: id,
			idFinLancamentoPagar: idFinLancamentoPagar,
			idFinChequeEmitido: idFinChequeEmitido,
			idFinStatusParcela: idFinStatusParcela,
			idFinTipoPagamento: idFinTipoPagamento,
			numeroParcela: numeroParcela,
			dataEmissao: dataEmissao,
			dataVencimento: dataVencimento,
			dataPagamento: dataPagamento,
			descontoAte: descontoAte,
			valor: valor,
			taxaJuro: taxaJuro,
			taxaMulta: taxaMulta,
			taxaDesconto: taxaDesconto,
			valorJuro: valorJuro,
			valorMulta: valorMulta,
			valorDesconto: valorDesconto,
			valorPago: valorPago,
			historico: historico,
		);			
	}

	
}